### DISCUSSION TITLE
mindconnect-nodejs cli - assign roles

### DISCUSSION THREAD

**Discussion by ValeWuerz** [mindsphere/mindconnect-nodejs#292] [Q&A] [answered]
Hello mindsphere developers.
Does anybody know how to assign roles to groups or users with the cli?
I can't find any commands that would serve this purpose.
Thanks a lot

**Comment by sn0wcat** [comment]
That doesn't work yet, because the Roles API is for some reason not an official API of MindSphere.

https://developer.mindsphere.io/apis/core-identitymanagement/api-identitymanagement-api-swagger-3-3-0.html

the roles are available on a special endpoint which is not documented : https://`{yourtenantname}`-settings.eu1.mindsphere.io/api/im/v3/myTenant/roles?page=0&size=50

and the policy for the CLI was to not use undocumented endpoints. // @ChristophDoehl

**Comment by sn0wcat** [comment, marked as answer]
But you can use the groups and membergroups interface to accomplish the same thing:

### 1. creating the group

the 4 default roles of mindsphere are  called 
```text
mdsp:core:Developer
mdsp:core:StandardUser
mdsp:core:TenantAdmin
mdsp:core:DeveloperAdmin
```

so you can do something like this:

```bash
 mc iam --mode create --group MyAdmins
``` 
which creates a new user group (note that the CLI will automatically prefix the group with `mdsp_usergroup:`

![image](https://user-images.githubusercontent.com/1426617/133111861-4bd3e5f1-4077-4d9b-9e33-19c38dd39541.png)

User Group in the MindSphere

![image](https://user-images.githubusercontent.com/1426617/133112162-cd95e958-95b4-4f43-8456-f61533b23a10.png)


After that you can assign the roles and subgroups via the --assign command

### 2) lets give the users in the group the Admin Role

```bash
 mc iam --mode assign  --membergroup mdsp_usergroup:MyAdmins --group mdsp:core:TenantAdmin --raw
``` 
which assigns the role TenantAdmin to the group MyAdmins

please notice the `--raw switch` which lets the user type in all the prefixes like mdsp_usergroup and mdsp:core:

*note: <strike>there is apparently a copy paste error in the order of the parameters so that the CLI gives a bit of misleading output </strike>, apparently, mindsphere swaps the order of parameter when assigning the roles when compared to child groups for some reason, we will fix that in the next version, just remember that for now the --membergroup is the group you are assigning roles to :)

![image](https://user-images.githubusercontent.com/1426617/133122609-a9b8a812-411c-4ecd-b8f1-70cc864e2a76.png)

The group has now admin role:

![image](https://user-images.githubusercontent.com/1426617/133122848-f82c79aa-389b-4160-830f-6f6de96ede1f.png)


### Lets give the group also vfc admin role.

```bash
mc iam --mode assign  --membergroup mdsp_usergroup:MyAdmins --group mdsp:core:vfc.admin --raw
```
![image](https://user-images.githubusercontent.com/1426617/133123110-7b571220-0269-4bfb-b856-ae02fe7b792b.png)

![image](https://user-images.githubusercontent.com/1426617/133123403-aa3515c5-3029-4c2e-9e41-d98194f4670f.png)


### 3) Assign the group to the user:

```bash
 mc iam --mode assign --user sn0wcat_at_mindsphere.io --group mdsp_usergroup:MyAdmins --raw
```
![image](https://user-images.githubusercontent.com/1426617/133123681-0da7f003-e893-437d-93e7-3e487a6fe027.png)

User assignement in mindsphere:

![image](https://user-images.githubusercontent.com/1426617/133123928-826f2c8b-2cb5-4f59-8a24-911b76735edd.png)

HTH
@ValeWuerz @ChristophDoehl

**Reply by ValeWuerz** [reply]
Perfect, that helps a lot. Thank you @sn0wcat.

**Comment by sn0wcat** [comment]
Update, there is now a role support in the CLI too.
